hal/src/thumbv7em/pukcc.rs

```diff
@@ -0,0 +1,579 @@
+            $(
+                .chain($data.iter().rev()) // > Little endian format! <
+            )+
+            .zip($crypto_ram.iter_mut())
+            .for_each(|(data_iter, cr_iter)| *cr_iter = *data_iter);
+
+            let mut _offset = 0;
+            $(
+                $name = &$crypto_ram[_offset.._offset + $data.len()];
+                _offset += $data.len();
+            )+
+            }
+        }
+}
+
+/// Struct representing a PUKCC peripheral.
+///
+/// It provides an access to cryptograhic algorithms in a safe, high-level
+/// manner
+pub struct Pukcc {
+    __: (),
+}
+
+impl Pukcc {
+    /// Constructor.
+    ///
+    /// Waits for a CryptoRAM readiness, enables a synchronous PUKCC clock and
+    /// performs a self test. In case a self test fails it returns an error
+    pub fn enable(mclk: &mut MCLK) -> Result<Self, SelfTestFailure> {
+        // TODO: Alleged RNG initialization. Passthrough constructor?
+        unsafe {
+            c_abi::wait_for_crypto_ram_clear_process();
+        }
+        mclk.ahbmask.modify(|_, w| w.pukcc_().set_bit());
+        let pukcc = Self { __: () };
+        pukcc.self_test().map(|_| pukcc)
+    }
+
+    /// Self test service.
+    ///
+    /// Clears up a CryptoRAM and does the checksum. If a checksum and a version
+    /// matches one defined in a HAL, it means that a self test passed
+    /// successfully.
+    ///
+    /// While using a high-level API, user should not need to use this service
+    /// explicitly.
+    pub fn self_test(&self) -> Result<(), SelfTestFailure> {
+        const PUKCL_VERSION: u4 = 0x04070100;
+        const CHECKNUM_1: u4 = 0x6E70DDD2;
+        const CHECKNUM_2: u4 = 0x25C8D64F;
+        let mut pukcl_params = c_abi::PukclParams::default();
+        unsafe {
+            c_abi::SelfTest::call(&mut pukcl_params);
+        }
+        let header = pukcl_params.header;
+        let service_params = unsafe { pukcl_params.params.SelfTest };
+        match header.u2Status.into() {
+            PukclReturnCode::Ok => {}
+            _ => return Err(SelfTestFailure(service_params)),
+        };
+        if service_params.u4Version != PUKCL_VERSION {
+            return Err(SelfTestFailure(service_params));
+        }
+        if service_params.u4CheckNum1 != CHECKNUM_1 {
+            return Err(SelfTestFailure(service_params));
+        }
+        if service_params.u4CheckNum2 != CHECKNUM_2 {
+            return Err(SelfTestFailure(service_params));
+        }
+
+        Ok(())
+    }
+
+    /// Service generating an ECDSA signature.
+    ///
+    /// GF(p) service. GF(2^n) variant is not implemented -- use low-level API.
+    ///
+    /// Input parameters:
+    /// - `hash`: `&[u8]` of length [`Curve::SCALAR_LENGTH`]
+    ///     - Hash of a message that is supposed to be signed.
+    /// - `private_key`: `&[u8]` of length [`Curve::SCALAR_LENGTH`]
+    ///     - Private key used for signing.
+    /// - `k`: `&[u8]` of length [`Curve::SCALAR_LENGTH`]
```

Comment:
It would be my personal preference if this wasnt a parameter and instead the module obtained entropy from some hardware source every time.

Other ideas:
 - Make a newtype `EntropySeed`, which is passed by value (so it is consumed). This newtype can be an enum with variants `FromHardware` (in which case the module obtains the entropy for you) or `InsecureFromSeed` which has the entropy contained within.
 - Separate `zp_ecdsa_sign` into two functions, one where the user provides the entropy (ie: `insecure_zp_ecdsa_sign_with_seed`) and the regular one which obtains the entropy for you (`zp_ecdsa_sign`)

Reply:
What does it mean it "obtains entropy for you"? I guess you mean you pass `impl RngCore` into the safe variant which then is used to populate a `k` slice locally.